src/number_theory/number_field/basic.lean

```diff
@@ -127,11 +130,81 @@ begin
     (((is_integral_closure.is_integral_algebra ℤ K).is_field_iff_is_field h_inj).mpr hf)
 end
 
-instance [number_field K] : is_dedekind_domain (𝓞 K) :=
+instance : is_dedekind_domain (𝓞 K) :=
 is_integral_closure.is_dedekind_domain ℤ ℚ K _
 
+variables (R : Type*) [comm_ring R] [algebra R K]
+
+lemma _root_.number_field.is_integral_closure.free (hc : is_integral_closure R ℤ K) :
+  free ℤ R :=
+begin
+  haveI : no_zero_smul_divisors ℤ R := function.injective.no_zero_smul_divisors _
+    hc.algebra_map_injective (map_zero _) (map_zsmul _),
+  haveI : is_noetherian ℤ R := is_integral_closure.is_noetherian _ ℚ K _,
+  have basis : Σ n, basis (fin n) ℤ R := free_of_finite_type_torsion_free',
+  obtain ⟨n, b⟩ := basis,
+  exact free.of_basis b,
+end
+
+lemma _root_.number_field.is_integral_closure.is_localization
+  (hc : is_integral_closure R ℤ K) :
+  is_localization (algebra.algebra_map_submonoid R (non_zero_divisors ℤ)) K :=
+begin
+  haveI : char_zero R := char_zero.of_module _ K,
+  haveI : is_domain R := ring_equiv.is_domain (𝓞 K) (number_field.ring_of_integers.equiv R).symm,
+  refine ⟨_, λ z, _, λ x y, ⟨λ h, ⟨1, _⟩, _⟩⟩,
+  { rintros ⟨_, ⟨x, ⟨hx, rfl⟩⟩⟩,
+    rw [is_unit_iff_ne_zero, map_ne_zero_iff _ (is_integral_closure.algebra_map_injective R ℤ K),
+      subtype.coe_mk, map_ne_zero_iff _ (ring_hom.injective_int (algebra_map ℤ R))],
+    exact mem_non_zero_divisors_iff_ne_zero.mp hx, },
+  { obtain ⟨m, hm⟩ := is_integral.exists_multiple_integral_of_is_localization
+      (non_zero_divisors ℤ) z (is_separable.is_integral ℚ z),
+    obtain ⟨x, hx⟩ := hc.is_integral_iff.mp hm,
+    exact ⟨⟨x, ⟨m, ⟨m, ⟨set_like.coe_mem m, rfl⟩⟩⟩⟩, by {simp only [hx, coe_coe,
+      set_like.coe_mk, map_int_cast, submonoid.smul_def, mul_comm, zsmul_eq_mul], }⟩, },
+  { simp only [hc.algebra_map_injective h], },
+  { rintros ⟨⟨_, ⟨m, ⟨hm, rfl⟩⟩⟩, h⟩,
+    refine congr_arg (algebra_map R K) ((mul_right_inj' _).mp h),
+    rw [subtype.coe_mk, map_ne_zero_iff _ (ring_hom.injective_int (algebra_map ℤ R))],
+    exact mem_non_zero_divisors_iff_ne_zero.mp hm, },
+end
+
+instance : free ℤ (𝓞 K) := is_integral_closure.free K _ infer_instance
+
+instance : is_localization (algebra.algebra_map_submonoid (𝓞 K) (non_zero_divisors ℤ)) K :=
+is_integral_closure.is_localization K _ infer_instance
+
+/-- A ℤ-basis of the ring of integers of `K`. -/
+noncomputable def basis : basis (free.choose_basis_index ℤ (𝓞 K)) ℤ (𝓞 K)
+:= free.choose_basis ℤ (𝓞 K)
+
 end ring_of_integers
 
+variable [number_field K]
+
+/-- A basis of `K` over `ℚ` that is also a basis of `𝓞 K` over `ℤ`. -/
+noncomputable def integral_basis : basis (free.choose_basis_index ℤ (𝓞 K)) ℚ K :=
+basis.localization_localization ℚ (non_zero_divisors ℤ) K (ring_of_integers.basis K)
```

Comment:
I think this would be the same as [`finite_dimensional.exists_is_basis_integral`](https://leanprover-community.github.io/mathlib_docs/ring_theory/dedekind_domain/integral_closure.html#finite_dimensional.exists_is_basis_integral) (except the definitions are nonconstructive so we can't compare them...)

Reply:
Yes, this is probably right. The construction I used has the benefit of giving an easy translation between basis of `K` and basis `𝓞 K`. That should also be possible using `finite_dimensional.exists_is_basis_integral` though. I could have a look at that if you mind code duplication.